exchanges/poloniex/poloniex_wrapper.go

```diff
@@ -491,259 +648,563 @@ func (e *Exchange) SubmitOrder(ctx context.Context, s *order.Submit) (*order.Sub
+	}
+	return err
 }
 
 // CancelBatchOrders cancels an orders by their corresponding ID numbers
 func (e *Exchange) CancelBatchOrders(ctx context.Context, o []order.Cancel) (*order.CancelBatchResponse, error) {
 	if len(o) == 0 {
-		return nil, order.ErrCancelOrderIsNil
+		return nil, common.ErrEmptyParams
+	}
+	type assetAndType struct {
+		aType asset.Item
+		oType order.Type
+		pair  currency.Pair // only futures orders need grouping by pair
 	}
-	orderIDs := make([]string, 0, len(o))
-	clientOrderIDs := make([]string, 0, len(o))
+	type orderAndClientSuppliedIDs struct {
+		orderIDs       []string
+		clientOrderIDs []string
+	}
+	assetAndTypeToIDsMap := make(map[assetAndType]*orderAndClientSuppliedIDs)
 	for i := range o {
-		switch {
-		case o[i].ClientOrderID != "":
-			clientOrderIDs = append(clientOrderIDs, o[i].ClientOrderID)
+		var pair currency.Pair
+		if o[i].AssetType == asset.Futures {
+			if o[i].Pair.IsEmpty() {
+				return nil, currency.ErrCurrencyPairEmpty
+			}
+			pair = o[i].Pair
+		} else if o[i].AssetType != asset.Spot {
+			return nil, fmt.Errorf("%w: %v", asset.ErrNotSupported, o[i].AssetType)
+		}
+		keyInfo := assetAndType{o[i].AssetType, o[i].Type, pair}
+		if assetAndTypeToIDsMap[keyInfo] == nil {
+			assetAndTypeToIDsMap[keyInfo] = &orderAndClientSuppliedIDs{
+				orderIDs:       []string{},
+				clientOrderIDs: []string{},
+			}
+		}
+		switch {
 		case o[i].OrderID != "":
-			orderIDs = append(orderIDs, o[i].OrderID)
+			assetAndTypeToIDsMap[keyInfo].orderIDs = append(assetAndTypeToIDsMap[keyInfo].orderIDs, o[i].OrderID)
+		case o[i].ClientOrderID != "":
+			assetAndTypeToIDsMap[keyInfo].clientOrderIDs = append(assetAndTypeToIDsMap[keyInfo].clientOrderIDs, o[i].ClientOrderID)
 		default:
 			return nil, order.ErrOrderIDNotSet
 		}
 	}
-	cancelledOrders, err := e.CancelMultipleOrdersByIDs(ctx, orderIDs, clientOrderIDs)
-	if err != nil {
-		return nil, err
-	}
 	resp := &order.CancelBatchResponse{
 		Status: make(map[string]string),
 	}
-	for i := range cancelledOrders {
-		if cancelledOrders[i].ClientOrderID != "" {
-			resp.Status[cancelledOrders[i].ClientOrderID] = cancelledOrders[i].State + " " + cancelledOrders[i].Message
-			continue
+	for key, value := range assetAndTypeToIDsMap {
+		if key.aType == asset.Spot {
+			switch key.oType {
+			case order.Market, order.Limit, order.AnyType, order.UnknownType:
+				cancelledOrders, err := e.CancelOrdersByIDs(ctx, value.orderIDs, value.clientOrderIDs)
+				if err != nil {
+					return nil, err
+				}
+				for _, co := range cancelledOrders {
+					if slices.Contains(value.orderIDs, co.OrderID) {
+						resp.Status[co.OrderID] = co.State + " " + co.Message
+					} else {
+						resp.Status[co.ClientOrderID] = co.State + " " + co.Message
+					}
+				}
+			case order.Stop, order.StopLimit, order.TrailingStop, order.TrailingStopLimit:
+				cancelledOrders, err := e.CancelMultipleSmartOrders(ctx, &CancelOrdersRequest{
```

Comment:
Has no test coverage
If it does, please stop using `require.NoError` and use `assert.NoError` for multi-case tests